--- /home/chrisn/src/ndpi-netfilter/nDPI//src/lib/third_party/include/ndpi_patricia.h	2022-06-30 09:34:28.890051275 -0400
+++ /home/chrisn/src/ndpi-netfilter/nDPI-patch//src/lib/third_party/include/ndpi_patricia.h	2022-06-30 13:30:03.254645035 -0400
@@ -49,13 +49,18 @@
 /* { from defs.h */
 #define ndpi_prefix_touchar(prefix) ((u_char *)&(prefix)->add.sin)
 
+#ifdef __KERNEL__
+#define MAXLINE 512
+#else
 #define MAXLINE 1024
+#endif
 
 #define BIT_TEST(f, b)  ((f) & (b))
 /* } */
 
 #define addroute make_and_lookup
 
+#ifndef __KERNEL__
 #include <sys/types.h> /* for u_* definitions (on FreeBSD 5) */
 #include <errno.h> /* for EAFNOSUPPORT */
 
@@ -75,7 +80,11 @@
 #include <winsock2.h>
 #include <ws2tcpip.h> /* IPv6 */
 #endif
-
+#else
+#include <linux/types.h>
+#include <linux/in.h>
+#include <linux/in6.h>
+#endif /* __KERNEL__ */
 /* { from mrt.h */
 
 typedef struct _prefix4_t {
