--- /home/chrisn/src/ndpi-netfilter/nDPI//src/include/ndpi_includes.h	2022-06-30 09:34:28.842051273 -0400
+++ /home/chrisn/src/ndpi-netfilter/nDPI-patch//src/include/ndpi_includes.h	2022-06-30 11:06:32.934283331 -0400
@@ -24,6 +24,7 @@
 #ifndef __NDPI_INCLUDES_H__
 #define __NDPI_INCLUDES_H__
 
+#ifndef __KERNEL__
 #include <stdint.h>
 #include <stdio.h>
 #include <stdarg.h>
@@ -33,10 +34,12 @@
 #include <time.h>
 #include <limits.h>
 #include <stdbool.h>
+#endif
 
 #ifdef WIN32
 #include "ndpi_win32.h"
 #else
+#ifndef __KERNEL__
 #include <sys/types.h>
 #include <sys/param.h>
 #include <arpa/inet.h>
@@ -46,10 +49,21 @@
 #include <netinet/ip.h>
 #include <netinet/tcp.h>
 #include <netinet/udp.h>
+#else
+#include <linux/ip.h>
+#include <linux/tcp.h>
+#include <linux/udp.h>
+#endif
 
 #if !defined __APPLE__ && !defined __FreeBSD__ && !defined __NetBSD__ && !defined __OpenBSD__
+
+#ifndef __KERNEL__
 #include <endian.h>
 #include <byteswap.h>
+#else
+#include <asm/byteorder.h>
+#include <linux/ctype.h>
+#endif
 
 #if defined __FreeBSD__ || defined __NetBSD__ || defined __OpenBSD__
 #include <netinet/in.h>
