--- /home/chrisn/src/ndpi-netfilter/nDPI//src/include/ndpi_typedefs.h	2022-06-30 09:34:28.846051273 -0400
+++ /home/chrisn/src/ndpi-netfilter/nDPI-patch//src/include/ndpi_typedefs.h	2022-06-30 11:21:30.962321055 -0400
@@ -956,8 +956,6 @@
 #define NUM_CUSTOM_CATEGORIES      5
 #define CUSTOM_CATEGORY_LABEL_LEN 32
 
-#ifdef NDPI_LIB_COMPILATION
-
 /* Needed to have access to HAVE_* defines */
 #ifndef _NDPI_CONFIG_H_
 #include "ndpi_config.h"
@@ -1100,8 +1098,6 @@
   struct ndpi_packet_struct packet;
 };
 
-#endif /* NDPI_LIB_COMPILATION */
-
 typedef enum {
    ndpi_cipher_safe = NDPI_CIPHER_SAFE,
    ndpi_cipher_weak = NDPI_CIPHER_WEAK,
