--- /home/chris/src/CyberReef/nDPI/src/include/ndpi_includes.h	2021-06-14 19:45:51.972249148 -0400
+++ ./ndpi_includes.h	2021-06-14 21:05:13.359341132 -0400
@@ -24,6 +24,7 @@
 #ifndef __NDPI_INCLUDES_H__
 #define __NDPI_INCLUDES_H__
 
+#ifndef __KERNEL__
 #include <stdint.h>
 #include <stdio.h>
 #include <stdarg.h>
@@ -32,10 +33,12 @@
 #include <ctype.h>
 #include <time.h>
 #include <limits.h>
+#endif
 
 #ifdef WIN32
 #include "ndpi_win32.h"
 #else
+#ifndef __KERNEL__
 #include <sys/types.h>
 #include <sys/param.h>
 #include <pthread.h>
@@ -46,10 +49,21 @@
 #include <netinet/ip.h>
 #include <netinet/tcp.h>
 #include <netinet/udp.h>
+#else
+#include <linux/ip.h>
+#include <linux/tcp.h>
+#include <linux/udp.h>
+#endif
 
 #if !defined __APPLE__ && !defined __FreeBSD__ && !defined __NetBSD__ && !defined __OpenBSD__
+
+#ifndef __KERNEL__
 #include <endian.h>
 #include <byteswap.h>
+#else
+#include <asm/byteorder.h>
+#include <linux/ctype.h>
+#endif
 
 #if defined __FreeBSD__ || defined __NetBSD__ || defined __OpenBSD__
 #include <netinet/in.h>
@@ -70,7 +84,7 @@
 #if defined __OpenBSD__
 #include "ndpi_includes_OpenBSD.h"
 #else
-typedef struct timeval pkt_timeval;
+typedef struct __kernel_old_timeval pkt_timeval;
 #endif /* __OpenBSD__ */
 
 #endif /* __NDPI_INCLUDES_H__ */
