--- /home/chrisn/src/ndpi-netfilter/nDPI//src/lib/third_party/src/ahocorasick.c	2022-06-30 09:34:28.890051275 -0400
+++ /home/chrisn/src/ndpi-netfilter/nDPI-patch//src/lib/third_party/src/ahocorasick.c	2022-07-01 11:29:20.182324150 -0400
@@ -43,6 +43,10 @@
 #include "ndpi_api.h"
 #include "ahocorasick.h"
 
+#ifdef __KERNEL__
+#define printf printk
+#endif
+
 /* TODO: For different depth of node, number of outgoing edges differs
    considerably, It is efficient to use different chunk size for 
    different depths */
