--- ../../../../../nDPI/src/lib/third_party/src/ht_hash.c	2021-06-29 15:18:26.064077493 +0000
+++ ./ht_hash.c	2021-07-07 18:50:38.196831331 +0000
@@ -1,11 +1,15 @@
 /* Based on https://gist.github.com/tonious/1377667 */
 
+#ifndef __KERNEL__
 #include <stdint.h>
 #include <sys/types.h>
 #include <stdlib.h>
 #include <stdio.h>
 #include <limits.h>
 #include <string.h>
+#else
+#include <linux/types.h>
+#endif
 
 #include "ht_hash.h"
 
@@ -24,7 +28,7 @@
 
   /* Allocate pointers to the head nodes. */
   if((hashtable->table = ndpi_malloc(sizeof(entry_t *) * size)) == NULL) {
-    free(hashtable);
+    ndpi_free(hashtable);
     return NULL;
   } else {    
     for(i = 0; i < size; i++)
@@ -63,7 +67,7 @@
     return NULL;  
   
   if((newpair->key = ndpi_strdup(key)) == NULL) {
-    free(newpair);
+    ndpi_free(newpair);
     return NULL;  
   }
 
@@ -82,7 +86,7 @@
   entry_t *last = NULL;
 
 #ifdef HASH_DEBUG
-  printf("*** %s() %s = %u ***\n", __FUNCTION__, key, value);
+  //printf("*** %s() %s = %u ***\n", __FUNCTION__, key, value);
 #endif
 
   bin = ht_hash(hashtable, key);
@@ -165,7 +169,7 @@
 }
 
 /* **************************************************** */
-
+#ifndef __KERNEL__
 #ifdef HASH_TEST
 
 int main(int argc, char **argv) {
@@ -185,3 +189,4 @@
 }
 
 #endif
+#endif
