--- /home/chrisn/src/3/ndpi-netfilter/nDPI//src/lib/third_party/src/node.c	2021-07-19 13:47:49.176005951 +0000
+++ /home/chrisn/src/3/ndpi-netfilter/nDPI-patch//src/lib/third_party/src/node.c	2021-07-19 14:54:09.380173153 +0000
@@ -17,10 +17,11 @@
  You should have received a copy of the GNU Lesser General Public License
  along with multifast.  If not, see <http://www.gnu.org/licenses/>.
 */
-
+#ifndef __KERNEL__
 #include <stdio.h>
 #include <string.h>
 #include <stdlib.h>
+#endif
 #include "ndpi_api.h"
 #include "../include/node.h"
 #include "sort.h"
@@ -76,7 +77,8 @@
 void node_release(AC_NODE_t * thiz, u_int8_t free_pattern)
 {
   if(free_pattern) {
-    for(int i=0; i<thiz->matched_patterns_num; i++) {
+    int i;
+    for(i=0; i<thiz->matched_patterns_num; i++) {
       if(!thiz->matched_patterns[i].is_existing)
         ndpi_free(thiz->matched_patterns[i].astring);
     }
