--- ../../../../../nDPI/src/lib/third_party/src/sha1-fast.c	2021-06-29 15:18:26.064077493 +0000
+++ ./sha1-fast.c	2021-06-29 19:45:08.676110777 +0000
@@ -20,9 +20,11 @@
  *   out of or in connection with the Software or the use or other dealings in the
  *   Software.
  */
-
+#ifndef __KERNEL__
 #include <stdint.h>
-
+#else
+#include <linux/types.h>
+#endif
 
 void sha1_compress(uint32_t state[5], const uint8_t block[64]) {
 #define ROTL32(x, n)  (((0U + (x)) << (n)) | ((x) >> (32 - (n))))  // Assumes that x is uint32_t and 0 < n < 32
