--- /home/chrisn/src/ndpi-netfilter/nDPI//src/lib/ndpi_classify.c	2022-06-30 09:34:28.858051274 -0400
+++ /home/chrisn/src/ndpi-netfilter/nDPI-patch//src/lib/ndpi_classify.c	2022-07-26 10:47:31.149448361 -0400
@@ -43,12 +43,14 @@
 
 #define _GNU_SOURCE
 
+#ifndef __KERNEL__
 #include <stdio.h>
 #include <ctype.h>
 // #include <sys/time.h>
 #include <stdlib.h>
 #include <stdint.h>
 #include <math.h>
+#endif
 #include "ndpi_main.h"
 #include "ndpi_classify.h"
 #include "ndpi_includes.h"
@@ -380,6 +382,7 @@
   }
 }
 
+#if 0
 /* transform times array to Markov chain */
 void
 ndpi_get_mc_rep_times (uint16_t *times, float *time_mc, uint16_t num_packets)
@@ -587,7 +590,7 @@
     break;
   }
 }
-
+#endif
 /* *********************************************************************
  * ---------------------------------------------------------------------
  *                      Time functions
@@ -682,10 +685,10 @@
 ndpi_log_timestamp(char *log_ts, uint32_t log_ts_len)
 {
   pkt_timeval tv;
+#if 0
   time_t nowtime;
   struct tm nowtm_r;
   char tmbuf[NDPI_TIMESTAMP_LEN];
-
   gettimeofday(&tv, NULL);
   nowtime = tv.tv_sec;
 #ifdef WIN32
@@ -697,4 +700,5 @@
 #endif
   strftime(tmbuf, NDPI_TIMESTAMP_LEN, "%H:%M:%S", &nowtm_r);
   snprintf(log_ts, log_ts_len, "%s.%06ld", tmbuf, (long)tv.tv_usec);
+#endif
 }
