--- /Users/asdasdsdsadsada/CLionProjects/ndpi-netfilter/nDPI//src/lib/protocols/bittorrent.c	2024-06-15 22:53:37
+++ /Users/asdasdsdsadsada/CLionProjects/ndpi-netfilter/nDPI-patch//src/lib/protocols/bittorrent.c	2024-06-15 22:54:24
@@ -72,8 +72,9 @@
 
       if(bt_magic)
 	bt_hash = &bt_magic[19];
-    } else
-      bt_hash = (const char*)&flow->packet.payload[28];
+    } else if(flow->packet.payload_packet_len > 28) {
+        bt_hash = (const char*)&flow->packet.payload[28];
+    }
 
     if(bt_hash && (flow->packet.payload_packet_len >= (20 + (bt_hash-(const char*)flow->packet.payload))))
       memcpy(flow->protos.bittorrent.hash, bt_hash, 20);
@@ -451,8 +452,11 @@
 	  } else if((v0_flags < 6 /* ST_NUM_STATES */) && (v0_extension < 3 /* EXT_NUM_EXT */)) {
 	    u_int32_t ts = ntohl(*((u_int32_t*)&(packet->payload[4])));
 	    u_int32_t now;
-
+#ifndef __KERNEL__
 	    now = (u_int32_t)time(NULL);
+#else
+	    now = (u_int32_t)ktime_get_real();
+#endif
 
 	    if((ts < (now+86400)) && (ts > (now-86400))) {
 	      bt_proto = ndpi_strnstr((const char *)&packet->payload[20], "BitTorrent protocol", packet->payload_packet_len-20);
