--- ../../../../../nDPI/src/lib/protocols/git.c	2021-06-14 12:55:42.806929729 -0400
+++ ./git.c	2021-06-16 12:42:31.663204060 -0400
@@ -21,13 +21,18 @@
 #include "ndpi_protocol_ids.h"
 
 #define NDPI_CURRENT_PROTO NDPI_PROTOCOL_GIT
-
+#ifndef __KERNEL__
 #include <stdlib.h>
+#endif
 #include "ndpi_api.h"
 
 
 #define GIT_PORT 9418
 
+#ifdef __KERNEL__
+extern int atoi(const char *str);
+#endif
+
 void ndpi_search_git(struct ndpi_detection_module_struct *ndpi_struct,
 		     struct ndpi_flow_struct *flow)
 {
