--- /home/chrisn/src/ndpi-netfilter/nDPI//src/lib/protocols/http.c	2022-06-30 09:34:28.874051274 -0400
+++ /home/chrisn/src/ndpi-netfilter/nDPI-patch//src/lib/protocols/http.c	2022-07-01 13:24:31.518614483 -0400
@@ -21,7 +21,9 @@
  *
  */
 
+#ifndef __KERNEL__
 #include <assert.h>
+#endif
 
 #include "ndpi_protocol_ids.h"
 
@@ -529,6 +531,15 @@
 }
 
 /* ************************************************************* */
+// http://lkml.iu.edu/hypermail/linux/kernel/0303.2/2007.html
+static unsigned int inet_addr(char *str)
+{
+  int a,b,c,d;
+  char arr[4];
+  sscanf(str,"%d.%d.%d.%d",&a,&b,&c,&d);
+  arr[0] = a; arr[1] = b; arr[2] = c; arr[3] = d;
+  return *(unsigned int*)arr;
+}
 
 static void ndpi_check_numeric_ip(struct ndpi_detection_module_struct *ndpi_struct,
 				  struct ndpi_flow_struct *flow,
@@ -543,7 +554,7 @@
     double_dot[0] = '\0';
 
   ip_addr.s_addr = inet_addr(buf);
-  if(strcmp(inet_ntoa(ip_addr), buf) == 0)
+  if(strcmp(ndpi_inet_ntoa(ip_addr), buf) == 0)
     ndpi_set_risk(ndpi_struct, flow, NDPI_HTTP_NUMERIC_IP_HOST);
 }
 
