--- /home/chrisn/src/3/ndpi-netfilter/nDPI//src/lib/protocols/tls.c	2021-07-19 13:47:49.176005951 +0000
+++ /home/chrisn/src/3/ndpi-netfilter/nDPI-patch//src/lib/protocols/tls.c	2021-07-19 15:26:20.240254265 +0000
@@ -29,6 +29,43 @@
 #include "ndpi_md5.h"
 #include "ndpi_sha1.h"
 
+#ifndef MIN
+#define MIN(X, Y) (((X) < (Y)) ? (X) : (Y))
+#endif
+
+#ifdef __KERNEL__
+// Algorithm: http://howardhinnant.github.io/date_algorithms.html
+static int days_from_civil(int y, int m, int d)
+{
+    y -= m <= 2;
+    int era = y / 400;
+    int yoe = y - era * 400;                                   // [0, 399]
+    int doy = (153 * (m + (m > 2 ? -3 : 9)) + 2) / 5 + d - 1;  // [0, 365]
+    int doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;           // [0, 146096]
+    return era * 146097 + doe - 719468;
+}
+
+static time_t timegm(struct tm* t)
+{
+    int year = t->tm_year + 1900;
+    int month = t->tm_mon;          // 0-11
+    if (month > 11)
+    {
+        year += month / 12;
+        month %= 12;
+    }
+    else if (month < 0)
+    {
+        int years_diff = (11 - month) / 12;
+        year -= years_diff;
+        month += 12 * years_diff;
+    }
+    int days_since_1970 = days_from_civil(year, month + 1, t->tm_mday);
+
+    return 60 * (60 * (24L * days_since_1970 + t->tm_hour) + t->tm_min) + t->tm_sec;
+}
+#endif
+
 extern char *strptime(const char *s, const char *format, struct tm *tm);
 extern int processClientServerHello(struct ndpi_detection_module_struct *ndpi_struct,
 				    struct ndpi_flow_struct *flow, uint32_t quic_version);
@@ -342,7 +379,7 @@
 
 	if(len < (sizeof(utcDate)-1)) {
 	  struct tm utc;
-	  utc.tm_isdst = -1; /* Not set by strptime */
+	  // utc.tm_isdst = -1; /* Not set by strptime */
 
 	  strncpy(utcDate, (const char*)&packet->payload[i+4], len);
 	  utcDate[len] = '\0';
@@ -376,7 +413,7 @@
 
 	    if(len < (sizeof(utcDate)-1)) {
 	      struct tm utc;
-	      utc.tm_isdst = -1; /* Not set by strptime */
+	      //utc.tm_isdst = -1; /* Not set by strptime */
 
 	      strncpy(utcDate, (const char*)&packet->payload[offset], len);
 	      utcDate[len] = '\0';
